File: Seori/LeetCode 75/Sliding Window/1004. Max Consecutive Ones III.py
class Solution:
    def longestOnes(self, nums: List[int], k: int) -> int:
        # [1] k는 sliding window 안에 존재할 수 있는 0의 개수를 의미한다.
        left, right = 0, 0
        for right in range(len(nums)):
            # [1-1] 0이 window에 들어올 때마다 k를 감소시킨다.
            if nums[right] == 0:
                k -= 1
            
            # [1-2] 0이 k보다 많이 들어왔다면 window의 길이가 늘어나지 않도록 left가 움직인다.
            if k < 0:
                if nums[left] == 0:
                    k += 1
                left += 1
        
        return right - left + 1

File: Gombang/[LeetCode75] - [Sliding Window]/1004_Max_Consecutive_Ones_III.cpp
// zeroCount가 k를 넘어가는 순간 left변수를 땡겨오는 방식으로 진행하였습니다.
class Solution {
public:
    int longestOnes(vector<int>& nums, int k) {
        int left = 0;
        int zeroCount = 0;
        int result = 0;

        for (int right = 0; right < nums.size(); right++)
        {
            if (nums[right] == 0)
                zeroCount++;

            while (zeroCount > k)
            {
                if (nums[left] == 0)
                    zeroCount--;

                left++;
            }

            result = max(result, right - left + 1);
        }

        return result;
    }
};